Auto-generate `shield()` calls using template literal types, #23

louwers opened the pull request:
As discussed, a Draft MR for #23 :+1: 

Action points:

* [x] Getting the contents of `types.ts` for the test and inside the transformer in a clean way (i.e. without using `__dirname`). (See `FIXME` inside `transform.ts` and `transformer.spec.ts`).
* [x] ~~I realized that I didn't consider optional tuple elements, so they are currently broken. This is a problem for optional function arguments:~~

```ts
export function doSomething(arg: string, arg2?: string)
```

* [ ] Another case I realized is tuples which contain undefined or null. `[string, null]`. Right now this is unrepresentable with `shield()`. Not such a big problem because who would want this. But it should just fail to generate a `shield()` call at all if someone tries.

brillout commented on examples/vite/package.json:
```diff
@@ -3,11 +3,12 @@
     "prod": "npm run build && npm run server:prod",
     "build": "vite build && vite build --ssr",
-    "server": "node server.js",
+    "server": "esno ./server",
```
We don't need `esno` since the server is still JS.

louwers replied:
True!

brillout commented on telefunc/node/vite/transform.ts:
```diff
@@ -2,19 +2,29 @@ import { Plugin } from 'vite'
     },
     async transform(src, id, options) {
       if (isSSR_options(options)) {
+        if (id.endsWith("telefunc.ts")) {
```
~~`telefunc.ts`~~ `.telefunc.ts`

brillout commented on telefunc/node/vite/transform.ts:
```diff
@@ -2,19 +2,29 @@ import { Plugin } from 'vite'
     async transform(src, id, options) {
       if (isSSR_options(options)) {
+        if (id.endsWith("telefunc.ts")) {
+          return shieldTelefunctions(src, typesSrc)
```
How about ~~`shieldTelefunctions()`~~ `generateShield()`?

brillout commented on telefunc/node/server/shield/codegen/transformer.ts:
```diff
@@ -0,0 +1,97 @@
+export { shieldTelefunctions }
+
+import { Project, VariableDeclarationKind } from 'ts-morph'
+
```
We can add
```js
const typesSrc = readFileSync(`${__dirname}/types.d.ts`)
```
 here.

`__dirname` doesn't exist in Cloudflare Workers but that's not a problem here.

This module is always loaded in Node.js (in dev and during the build process).

In production this module is not loaded. (None of Vite's code is actually loaded in production.)

louwers replied:
That doesn't work during test.

Then `__dirname` is:

```
telefunc/telefunc/dist/tests/server/shield/codegen/types.d.ts
```

brillout replied:
> `types.d.ts` is not present at all

```ts
const telefuncSrcDir = `${__dirname}/../../../../telefunc`
const typesDeclarationFile = `${telefuncSrcDir}/node/server/shield/codegen/types.d.ts`
```
While adding `node/server/shield/codegen/types.d.ts` to `telefunc/pacakge.json#files`.

Or we configure TS to add `types.d.ts` in `dist/` (not sure if that's possible).


> `.d.ts` files don't contain type level types

Not sure what you mean.

louwers replied:
That works!

This is what I meant: 

`dist/tests`:
![image](https://user-images.githubusercontent.com/649392/164058415-7ccc7c6d-ce06-4a55-adc0-4b420747aa26.png)

`dist/node`:

![image](https://user-images.githubusercontent.com/649392/164058485-5b3a887c-1321-476c-9faf-80adf9b191e7.png)

brillout commented on telefunc/node/server/shield/codegen/transformer.ts:
```diff
@@ -0,0 +1,97 @@
+  for (let message of messages) {
+    const found = message.match(matcher)
+    if (!found) {
+      console.warn("Encountered unrecognized emit message: ", message)
```
While this error message and the one below are convenient for us, they are confusing for users. We should use `assert()` instead. We can still add some debug info if we want. Example of using `assert()` with some debug info:

https://github.com/vikejs/telefunc/blob/4aa22608b5be742a6b1c5ff78c8535a41aa32547/telefunc/node/vite/loadTelefuncFilesWithVite.ts#L22

louwers replied:
Hmm, I don't want to stop on failure. Just a warning would suffice.

brillout commented on telefunc/node/server/shield/codegen/transformer.ts:
```diff
@@ -0,0 +1,97 @@
+  const messages = diagnostics.map(d => d.getMessageText().toString())
+  const teleFunToShieldString = parseEmitMessages(messages)
+
+  // TODO: do users ever want to add shield() calls themselves?
```
Alternatively we can just add a collision avoiding prefix, so that the generated code is:

```js
import { shield as __shieldGenerator_shield } from 'telefunc'
const __shieldGenerator_t = __shieldGenerator_shield.type
````

brillout replied:
>   // TODO: do users ever want to add shield() calls themselves?

I'm also just realizing that not supporting this is a breaking change. But we can bump the semver major if needs be.

bartenra replied:
What happens when calling `shield` multiple times for the same function?

brillout replied:
Not sure. I'll have to check. But we can show a warning, I guess that'd be enough to consider this not a breaking change.

Also one thing we can do is to flag the `shield()` call: `shield(fn, [t.string], { __generated: true })`. So that we can let the user's shield take over.

louwers replied:
I made it backward-compatible.

brillout commented on telefunc/node/server/shield/codegen/transformer.ts:
```diff
@@ -0,0 +1,97 @@
+  for (const teleFunName of teleFunNames) {
+    shieldStrSourceFile.addVariableStatement({
+      declarations: [{
+        name: `${teleFunName}Shield`,
```
Same here, we should use a collision avoiding prefix.

brillout commented:
Performance shouldn't be a problem, but let's have an idea of how long the transpilation is delayed by the new transformer.

Worst case we only apply the transformer for prod building. Since the user's IDE already does the type checking job in dev. We can always make this an option.

louwers commented:
W.r.t performance I generated this file: https://gist.github.com/louwers/392db2e700f9f597d8eecc24b7836c1f

Good news, transformer disabled:

```
real    0m0,881s
user    0m0,806s
sys     0m0,085s
```

Enabled:

```
real    0m2,883s
user    0m4,400s
sys     0m0,159s
```

Which is very good for such a complex types and so many functions. So I'd say we just keep it enabled for dev mode, but monitor real-world performance.

Bad news: I found that the diagnostics I use for extracting the value of template literal types gets cut off:

```
Encountered unrecognized emit message:  Type '"fun94"' is not assignable to type '"[t.union(t.const(94), { geometries: t.array(t.union({ coordinates: t.array(t.array(t.array(t.array(t.number)))), type: t.const('MultiPolygon') }, { coordinates: t.array(t.array(t.array(t.number))), type: t.const('Polygon') }, { type: t.const('MultiLineString'), coordinates: t.array(t.array(t.array(t.number))) }, { ...'.
```
so I have to find a better approach. The author of ts-morph is likely on it as well: https://github.com/dsherret/ts-morph/issues/1266

brillout commented:
> I'd say we just keep it enabled for dev mode, but monitor real-world performance.

👍 We may want to have development experience that is as fast as possible development, so yea let's monitor on real world apps.

> Bad news: I found that the diagnostics I use for extracting the value of template literal types gets cut off:

Seems like it would happen only rarely and, more importantly, deterministically though.

brillout commented:
Btw. a potential solution for the problem of the emit info being truncated:

I can see that the threshold for truncating to be some `const` variable in TypeScript.

If that's the case then it's easy to change it to some big number `9999` in a post-install script. (I believe TypeScript ships as JavaScript and not binary; should be easy to patch TypeScript.)

We pin TypeScript and check if the patch still works when we use a newer TypeScript version.

brillout commented on telefunc/tsconfig.tests.json:
```diff
@@ -1,6 +1,7 @@
 {
   "extends": "./tsconfig.base.json",
   "include": ["./**/*.spec.ts"],
+  "files": ["node/server/shield/codegen/types.ts"],
```
Any reason we don't add it at `telefunc/node/tsconfig.json` instead?

louwers replied:
It's already included there.

> Btw. a potential solution for the problem of the emit info being truncated:
> 
> I can see that the threshold for truncating to be some `const` variable in TypeScript.
> 
> If that's the case then it's easy to change it to some big number `9999` in a post-install script. (I believe TypeScript ships as JavaScript and not binary; should be easy to patch TypeScript.)
> 
> We pin TypeScript and check if the patch still works when we use a newer TypeScript version.

I'm using `.getLiteralValue()` in a way that seems to work consistently now!

louwers replied:
That `tsconfig.json` does not apply to the tests.

brillout replied:
Before running the tests all TS files are transpiled. So I believe it should work to add it at `telefunc/node/tsconfig.json` instead. It would be cleaner (`types.ts` is an integrated part of the `node/` build; it's not only used for tests).

louwers replied:
Yes but they are not transpiles to the `tests` folder. Right now `types.d.ts` can be imported with `${__dirname}/types.d.ts` and something can be said for that as well. Feel free to change it though.

brillout replied:
You're right, my bad.

louwers commented:
> Btw. a potential solution for the problem of the emit info being truncated:
> 
> I can see that the threshold for truncating to be some `const` variable in TypeScript.
> 
> If that's the case then it's easy to change it to some big number `9999` in a post-install script. (I believe TypeScript ships as JavaScript and not binary; should be easy to patch TypeScript.)
> 
> We pin TypeScript and check if the patch still works when we use a newer TypeScript version.

I'm using `.getLiteralValue()` in a way that seems to work consistently now!

brillout commented on telefunc/node/server/shield/codegen/transformer.ts:
```diff
@@ -92,9 +66,13 @@ const generateShield = (
+    const shieldStr = typeAlias.getType().getLiteralValue()
+
+    if (!shieldStr || typeof shieldStr !== 'string') {
+      assertWarning(true, `Failed to generate shield() call for telefunction '${teleFunName}'`)
```
~~`assertWarning(true`~~ `assertWarning(false`

brillout replied:
Failed to generate shield() ~~call~~ for

brillout commented on telefunc/node/server/shield/codegen/transformer.ts:
```diff
@@ -0,0 +1,83 @@
+      assertWarning(false, `Failed to generate shield() for telefunction '${teleFunName}'`)
+      continue
+    }
+    const shieldStrWithAlias = shieldStr.replace(/t\./g, `${tAlias}.`)
```
```ts
function hello(name: 'some-t.string') {
}
```
I believe this would break?

louwers replied:
Oops.

brillout commented:
Let me know when the problem `function hello(name: 'some-t.string')` is solved. I'll then merge. FYI: I renamed variables to be consistent with the rest of Telefunc's source, and added an e2e test to `examples/vite/`.

louwers commented:
> Let me know when the problem `function hello(name: 'some-t.string')` is solved. I'll then merge. FYI: I renamed variables to be consistent with the rest of Telefunc's source, and added an e2e test to `examples/vite/`.

Awesome! :tada:  I'll let you know.

louwers commented:
Made an additional fix to make sure single quotation marks are escaped.

brillout commented:
Released in `0.1.20`.
